[FIX] account: prevent 'In Process' state if payment can't reconcile

mathcoutant opened the pull request:
### Steps to reproduce:
- Accounting > Journals > Bank
- Set the journal "Bank" as "Outstanding Receipts Account" on the line "Manual Payment" in the page "Incoming Payments"
- Create a payment with a partner
- Confirm it, its state should be "Paid"
- Create an invoice with the same partner and amount
- Confirm and click the add button to link the payment
- Return on the payment, its state is back to "In Process"

### Cause:
This [commit](https://github.com/odoo/odoo/commit/533984ac5c10fbd91742f34b740c63c04fbef094) added the return to the state "In process" when payment are unreconciled. The check is:
`if move.currency_id.is_zero(sum(liquidity.mapped('amount_residual')))`
 
When setting the journal "Bank" as the outstanding account, the line which appear in `liquidity` has the payment amount in `amount_currency` and `amount_residual`. So the amount is not 0 and the payment state is set back to "In Process".

### Solution:
When the account of the payment does not allow reconciliation, the payment is never supposed to be in the state "In Process". To prevent this, this commit adds a check on `account_id.reconcile`.

opw-4718717

yosuanicolaus reviewed (approved):
LGTM, thanks

Sending this to final review

antoine162 commented on addons/account/tests/test_account_move_reconcile.py:
```diff
@@ -5266,3 +5266,28 @@ def reconcile_move(move, transaction_amount, balance=None, date='2023-09-30', cu
+
+        lines_to_reconcile = (invoice + payment.move_id).line_ids.filtered(lambda x: x.account_id.account_type == 'asset_receivable')
+        lines_to_reconcile.reconcile()
+        payment._compute_state()
```
Instead of calling `_compute_state` directly, it would be better to unreconcile the lines to be closer to the actual flow. You'll probably need to call `_flush_recordset` in order to trigger the compute in the test (just add a comment explaining that).

mathcoutant replied:
I tried to unreconcile and flush as you said, but was not triggering `_compute_state` for some reason.
While testing some things, I noticed that the invoice/reconciliation is not even needed to trigger the bug. Calling `_compute_state` on the paid payment will set it back to "In progress".

I updated my test with this behavior. Tell me if you think it's ok

antoine162 replied:
Personally, I would prefer to not call `_compute_state` directly in the test, but find a way to trigger the compute.
This is because in real life compute methods on stored fields are called in a very particular context (during flush, with some fields protected, and with other fields computed before/after). It's best for the test to replicate that.

It's very weird that you wouldn't be able to trigger the compute in the test.

antoine162 reviewed (commented):
I would suggest changing
https://github.com/odoo/odoo/blob/3602f236c61dcf41a901824d4ba79de9fd894397/addons/account/models/account_payment.py#L1065
to use the same condition (i.e. `account.reconcile == False`)

Maybe this logic can even be factorized between `_action_post` and `_compute_state`.

@william-andre I'd appreciate your thoughts on that as well as any other comments you might have.

It would be better for the test to be in `test_account_payment.py` right next to the test from https://github.com/odoo/odoo/pull/190225

william-andre reviewed (commented):
Current fix is fine, but could you fix the style error?

william-andre commented on addons/account/tests/test_account_payment.py:
```diff
@@ -652,6 +652,27 @@ def test_payment_confirmation_with_bank_outstanding_account(self):
+        self.company_data['default_journal_bank'].outbound_payment_method_line_ids[0].payment_account_id = unreconciliable_account
+        invoice = self.init_invoice(move_type='out_invoice', amounts=[10], post=True)
+
+        payment = self.env['account.payment.register']\
+            .with_context(active_model='account.move', active_ids=invoice.ids)\
+            .create({
+                'amount': 10.0, 
+                'journal_id': self.company_data['default_journal_bank'].id,
+                'payment_method_line_id': self.company_data['default_journal_bank'].outbound_payment_method_line_ids[0].id,
+                'currency_id': self.company_data['currency'].id})\
+            ._create_payments()
```
```suggestion
        payment = self.env['account.payment.register'].with_context(
            active_model='account.move', 
            active_ids=invoice.ids,
        ).create({
            'amount': 10.0,
            'journal_id': self.company_data['default_journal_bank'].id,
            'payment_method_line_id': self.company_data['default_journal_bank'].outbound_payment_method_line_ids[0].id,
            'currency_id': self.company_data['currency'].id,
        })._create_payments()
```

Also, avoid `\` to continue the line, use parenthesis instead

https://peps.python.org/pep-0008/#maximum-line-length
> The preferred way of wrapping long lines is by using Python’s implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.

mathcoutant commented:
@antoine162 I adapted the test and fixed the style issue.
The change you suggested is making a lot of tests fail (I guess the tests were not well configured) so I removed it.

Is it ok to be approved?

smetl commented on addons/account/models/account_payment.py:
```diff
@@ -416,7 +416,7 @@ def _compute_state(self):
                 payment.state = (
                     'paid'
-                    if move.currency_id.is_zero(sum(liquidity.mapped('amount_residual'))) else
+                    if move.currency_id.is_zero(sum(liquidity.mapped('amount_residual'))) or not liquidity.account_id.reconcile else
```
Fun part, there is another bug here:
```suggestion
                    if move.currency_id.is_zero(sum(liquidity.mapped('amount_residual_currency'))) or not liquidity.account_id.reconcile else
```
or
```suggestion
                    if move.company_currency_id.is_zero(sum(liquidity.mapped('amount_residual'))) or not liquidity.account_id.reconcile else
```

smetl commented on addons/account/tests/test_account_payment.py:
```diff
@@ -652,6 +652,28 @@ def test_payment_confirmation_with_bank_outstanding_account(self):
         self.assertEqual(payment.state, 'paid')
 
+    def test_payment_state_with_unreconciliable_outstanding_account(self):
+        unreconciliable_account = self.env['account.account'].create({
+            'code': '209.01.01',
+            'name': 'Bank Account',
+            'account_type': 'asset_cash',
+            'reconcile': False,
+        })
+        self.company_data['default_journal_bank'].outbound_payment_method_line_ids[0].payment_account_id = unreconciliable_account
+        invoice = self.init_invoice(move_type='out_invoice', amounts=[10], post=True)
+
+        payment = self.env['account.payment.register'].with_context(
+            active_model='account.move',
+            active_ids=invoice.ids,
+        ).create({
+            'amount': 10.0,
+            'journal_id': self.company_data['default_journal_bank'].id,
+            'payment_method_line_id': self.company_data['default_journal_bank'].outbound_payment_method_line_ids[0].id,
+            'currency_id': self.company_data['currency'].id,
+        })._create_payments()
```
The register payment wizard already computes a lot of default values
```suggestion
        unreconciliable_account = self.env['account.account'].create({
            'code': '209.01.01',
            'name': 'Bank Account',
            'account_type': 'asset_cash',
            'reconcile': False,
        })
        self.company_data['default_journal_bank'].outbound_payment_method_line_ids.payment_account_id = unreconciliable_account
        invoice = self.init_invoice(move_type='out_invoice', amounts=[10], post=True)

        payment = self.env['account.payment.register'].with_context(
            active_model='account.move',
            active_ids=invoice.ids,
        ).create({})._create_payments()
```

mathcoutant replied:
I need to at least keep the line on `'payment_method_line_id'` for the test to work

william-andre reviewed (approved):
@robodoo r+
